ts/components/loader.ts

```diff
@@ -65,10 +73,53 @@ export interface MathJaxObject extends MJObject {
     preLoad: (...names: string[]) => void;            // Indicate that packages are already loaded by hand
     defaultReady: () => void;                         // The function performed when all packages are loaded
     getRoot: () => string;                            // Find the root URL for the MathJax files
+    pathFilters: PathFilterList;                      // the filters to use for looking for package paths
   };
   startup?: any;
 }
 
+/**
+ * Functions to used to filter the path to a package
+ */
+export const PathFilters: {[name: string]: PathFilterFunction} = {
+  /**
+   * Look up the path in the configuration's source list
+   */
+  source: (data) => {
+    if (CONFIG.source.hasOwnProperty(data.name)) {
+      data.name = CONFIG.source[data.name];
+    }
+    return true;
+  },
+
+  /**
+   * Add [mathjax] before any relative path, and add .js if needed
+   */
+  normalize: (data) => {
+    const name = data.name;
+    if (!name.match(/^(?:[a-z]+:\/)?\/|[a-z]:\\|\[/i)) {
+      data.name = '[mathjax]/' + name.replace(/^\.\//, '');
+    }
+    if (data.addExtension && !name.match(/\.[^\/]+$/)) {
+      data.name += '.js';
+    }
+    return true;
+  },
+
+  /**
+   * Revursively replace path prefixes (e.g., [mathjax], [tex], etc.)
+   */
+  prefix: (data) => {
+    let match;
+    while ((match = data.name.match(/^\[([^\]]*)\]/))) {
+      if (!CONFIG.paths.hasOwnProperty(match[1])) break;
```

Comment:
So what if the the path that is used actually has a `[path]` part and components need to be replace _after_ that?
We simply don't allow that?

Reply:
I'm sorry, I'm not sure I understand the question.

Because this is a FunctionList, which is a PrioritizedList, you can add filters into the list at any location, so you can have a filter that adjusted the path after `[path]` has already been resolved, or you could add a filter that modifies the path before that.  It would depend on the need.  Prior to this PR we don't have any means of doing that (other than using the `source` configuration to directly map a request to its location).

I'm sure I'm not getting what you are driving at.

Reply:
For the record, we discussed this face-to-face, and I now understand that you were asking about an actual file name that contains `[...]`.  Since the substitutions are only made when (a) the `[...]` is at the start of the path name, and (b) only when there is an actual path for that prefix defined in the `paths` object, it is unlikely that this will cause a problem, and if it did, one could always use `./[prefix]...` to prevent this filter from modifying it.